plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/TestMariaDbTypeMapping.java

```diff
@@ -876,6 +835,100 @@ public void testTimestampCoercion()
+
+    private void testDatetime(ZoneId sessionZone)
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(getTimeZoneKey(sessionZone.getId()))
+                .build();
+        createCommonDatetimeTimestampRoundTrip("datetime")
+                // before epoch
+                .addRoundTrip("datetime(3)", "TIMESTAMP '1958-01-01 13:18:03.123'", createTimestampType(3), "TIMESTAMP '1958-01-01 13:18:03.123'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1969-12-31 23:59:59.999995'", createTimestampType(6), "TIMESTAMP '1969-12-31 23:59:59.999995'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1969-12-31 23:59:59.999949'", createTimestampType(6), "TIMESTAMP '1969-12-31 23:59:59.999949'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '1969-12-31 23:59:59.999994'", createTimestampType(6), "TIMESTAMP '1969-12-31 23:59:59.999994'")
+
+                //After epoch
+                .addRoundTrip("datetime(3)", "TIMESTAMP '3555-01-01 13:18:03.123'", createTimestampType(3), "TIMESTAMP '3555-01-01 13:18:03.123'")
+                .addRoundTrip("datetime(6)", "TIMESTAMP '5555-12-31 23:59:59.999995'", createTimestampType(6), "TIMESTAMP '5555-12-31 23:59:59.999995'")
+
+                // null
+                .addRoundTrip("datetime(0)", "NULL", createTimestampType(0), "CAST(NULL AS TIMESTAMP(0))")
+                .addRoundTrip("datetime(1)", "NULL", createTimestampType(1), "CAST(NULL AS TIMESTAMP(1))")
+                .addRoundTrip("datetime(2)", "NULL", createTimestampType(2), "CAST(NULL AS TIMESTAMP(2))")
+                .addRoundTrip("datetime(3)", "NULL", createTimestampType(3), "CAST(NULL AS TIMESTAMP(3))")
+                .addRoundTrip("datetime(4)", "NULL", createTimestampType(4), "CAST(NULL AS TIMESTAMP(4))")
+                .addRoundTrip("datetime(5)", "NULL", createTimestampType(5), "CAST(NULL AS TIMESTAMP(5))")
+                .addRoundTrip("datetime(6)", "NULL", createTimestampType(6), "CAST(NULL AS TIMESTAMP(6))")
+                .execute(getQueryRunner(), session, mariaDbCreateAndInsert("tpch.test_datetime"));
+    }
+
+    @Test
+    public void testIncorrectDateTime()
+    {
+        //MariaDB support any valid TIMESTAMP literal of type YYYY-MM-DD hh:mm:ss.S(up to 6 digits), fails otherwise.
+        try (TestTable table = new TestTable(server::execute, "test_incorrect_datetime", "(dt DATETIME)")) {
+            assertQueryFails(format("INSERT INTO %s VALUES (TIMESTAMP '999-01-01 00:00:00.000')", table.getName()), ".*not a valid TIMESTAMP literal.*");
+            assertQueryFails(format("INSERT INTO %s VALUES (TIMESTAMP '10000-01-19 03:14:08.000')", table.getName()), ".*Failed to insert data.*");
+        }
+    }
+
+    private SqlDataTypeTest createCommonDatetimeTimestampRoundTrip(String inputType)
+    {
+        String withPrecision3 = getInputTypeWithPrecision(inputType, 3);
+        String withPrecision6 = getInputTypeWithPrecision(inputType, 6);
```

Comment:
`inputType + "(3)"` & `inputType + "(6)"` look sufficient. I would inline these variables for consistency with other roundtrips.